### ISSUE TITLE
[FEATURE REQ] Cache key vault secrets

### ISSUE THREAD

**Issue by madkroll** [Azure/azure-sdk-for-java#19265]
**Is your feature request related to a problem? Please describe.**
Application uses KeyVault as a safe for secrets via **azure-security-keyvault-secrets**. However, sometimes it may experience connectivity issues and is not able to access key vault for short period of time (up to 5 minutes). Apart from setting up retry policies - it would be also useful to have a caching provided by SDK. In our case we are OK to cache a value for 10-30 minutes but at the same time be able to reset it in key vault, so once cache expired - a new value is taken.

**Describe the solution you'd like**
Provide caching layer to azure-security-keyvault-secrets in similar manner as it's done for .Net library:
https://docs.microsoft.com/en-us/samples/azure/azure-sdk-for-net/azure-key-vault-proxy/

So client code can specify the duration how long the particular secret should be cached for.

**Describe alternatives you've considered**
Alternatively we need to setup custom cache solution in our app (sure we also try to get rid of connectivity drops)

**Additional context**
-

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ + ] Description Added
- [ + ] Expected solution specified

**Comment by joshfree** [comment]
Thanks for filing this request @madkroll.  @vcolin7 can you please get this scheduled into a monthly milestone with the keyvault team

**Comment by vcolin7** [comment]
Sure @joshfree, I will bring it up with the team :)

**Comment by heaths** [comment]
We've previously discussed this with the service team, and feel that caching is an application concern. A web site could cache in many different ways that would different from console or window application (e.g. memory, file system, database, redis, etc.). Given we're a lower-level SDK, caching is a higher-level application concern with various different requirements. I've also brought this up with our architects and have discussed the possibility of a proxy-like built-in policy for this and other SDKs (e.g. AppConfig) one could opt into. This certainly adds weight to that request and I'll mention it.

The sample I wrote that's linked in the description was the answer to that, and maybe a sample in each language would be good. This way, with very little effort to plug into our SDKs' extensible pipelines, applications can cache with whatever requirements they have to wherever they need.

**Comment by madkroll** [comment]
Thanks Heath for your response.
To be honest, my first impression on that code sample was - it is an out-of-the-box solution from .Net SDK. I completely missed this is a sample code :D

Writing our custom policies is OK solution. It just requires to learn a bit deeper on how SDK library uses these policies internally.

**Comment by madkroll** [comment]
So for all mates who came to this ticket, check out original sources for code sample referenced in description (remember, it's for .Net SDK) and setup your custom way of caching by following same principles.

Here is a reference on source repo (might be moved in future, but you know where to search for it):
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/samples/keyvaultproxy/src/KeyVaultProxy.cs

Basically using AtomicReference as a value container in place of Cache should do a trick I believe.
